### PR TITLE
Trac3109 - Re-factor of EfvAttribute class needed - split into 2 classes

### PR DESCRIPTION
**Pull request by rpetry**
@alf239 - could you review please

### CODE REVIEW

**Comment by alf239** [comment]
It's too late, perhaps, but:
1. `EfoAttribute` has got no `serialVersionUID`;
2. `EfoAttribute.log` is an object field which is not transient. Not saved, I hope—but at least attempted to be serialized.

File: atlas-web/src/main/java/uk/ac/ebi/gxa/requesthandlers/query/ExperimentsPopupRequestHandler.java [hunk 1]
```diff
@@ -107,6 +108,8 @@ public class ExperimentsPopupRequestHandler extends AbstractRestRequestHandler {
                 if (term != null) {
                     jsResult.put("efv", term.getTerm());
                 }
+            } else if (Strings.isNullOrEmpty(factorValue)) {
+                attr = new EfAttribute(factor);
```
Comment (alf239) [inline atlas-web/src/main/java/uk/ac/ebi/gxa/requesthandlers/query/ExperimentsPopupRequestHandler.java:112]: We have the same logic in `EfvAttribute.encodePair`—does it mean that sometimes we care about the difference between EF and EFV and sometimes we don't?

If it doesn't, I'd suggest cleaning up `encodePair` before we merge.

File: atlas-web/src/main/java/ae3/service/AtlasBitIndexQueryService.java [hunk 2]
```diff
@@ -350,8 +350,8 @@ public class AtlasBitIndexQueryService implements AtlasStatisticsQueryService {
         long timeStart = System.currentTimeMillis();
         List<String> scoringEfs = new ArrayList<String>();
         if (bioEntityId != null) {
-            Set<EfvAttribute> scoringEfAttrs = statisticsStorage.getScoringEfAttributesForBioEntity(bioEntityId, statType);
-            for (EfvAttribute efAttr : scoringEfAttrs) {
+            Set<EfAttribute> scoringEfAttrs = statisticsStorage.getScoringEfAttributesForBioEntity(bioEntityId, statType);
+            for (EfAttribute efAttr : scoringEfAttrs) {
                 if (efAttr != null && (ef == null || "".equals(ef) || ef.equals(efAttr.getEf()))) {
                     scoringEfs.add(efAttr.getEf());
```
Comment (alf239) [inline atlas-web/src/main/java/ae3/service/AtlasBitIndexQueryService.java]: If you take a look at the usages of the code in question, you'll find out that each time we're creating a new `EfAttribute` for this `String`. So the question is, do we still need to return `List<String>` and not `List<EfAttribute>`?

```
Method
    getScoringEfsForBioEntity(Integer bioEntityId, StatisticsType statType, String ef):List<String> of interface ae3.service.AtlasStatisticsQueryService
Found usages  (4 usages)
    Production  (2 usages)
        atlas-web  (2 usages)
            ae3.model  (1 usage)
                AtlasGene  (1 usage)
                    getDifferentiallyExpressedFactors(Collection<String>, String, AtlasStatisticsQueryService)  (1 usage)
                        (468: 56) List<String> efs = atlasStatisticsQueryService.getScoringEfsForBioEntity(getGeneId(), UP_DOWN, ef);
            ae3.service  (1 usage)
                AtlasBitIndexQueryService  (1 usage)
                    getExperimentsSortedByPvalueTRank(Integer, Attribute, int, int, StatisticsType)  (1 usage)
                        (296: 32) List<String> efs = getScoringEfsForBioEntity(bioEntityId, statType, null);
```

**Comment by rpetry** [comment]
EfoAttribute is not stored in bitindex afaik - thus doesn't need serialVersionUID

File: atlas-web/src/main/java/uk/ac/ebi/gxa/requesthandlers/query/ExperimentsPopupRequestHandler.java (hunk 1 above)
Comment (rpetry) [inline atlas-web/src/main/java/uk/ac/ebi/gxa/requesthandlers/query/ExperimentsPopupRequestHandler.java:112]: Done

**Comment by alf239** [comment]
This pull request cannot be automatically merged — @rpetry, could you please check?

**Comment by rpetry** [comment]
It can now - thanks

File: atlas-web/src/main/java/ae3/model/AtlasGene.java [hunk 3]
```diff
@@ -465,15 +465,14 @@ public class AtlasGene {
             @Nullable String ef,
             AtlasStatisticsQueryService atlasStatisticsQueryService) {
         List<ExperimentalFactor> result = new ArrayList<ExperimentalFactor>();
-        List<String> efs = atlasStatisticsQueryService.getScoringEfsForBioEntity(getGeneId(), UP_DOWN, ef);
+        List<EfAttribute> efs = atlasStatisticsQueryService.getScoringEfsForBioEntity(getGeneId(), UP_DOWN, ef);
         efs.removeAll(omittedEfs);
```
Comment (alf239) [inline atlas-web/src/main/java/ae3/model/AtlasGene.java]: # No relationship between generic parameter and method argument

This call to a generic collection method `removeAll()` contains an argument with an incompatible class from that of the collection's parameter (i.e., the type of the argument is neither a supertype nor a subtype of the corresponding generic type argument). Therefore, it is unlikely that the collection contains any objects that are equal to the method argument used here. Most likely, the wrong value is being passed to the method.

Comment (rpetry) [inline atlas-web/src/main/java/ae3/model/AtlasGene.java]: Classic oversight... - well-spotted - my bad. Fixed.

File: atlas-web/src/main/java/ae3/model/AtlasGene.java [hunk 4]
```diff
@@ -461,19 +462,24 @@ public class AtlasGene {
...
-            Collection<String> omittedEfs,
+            final Collection<String> omittedEfs,
             @Nullable String ef,
             AtlasStatisticsQueryService atlasStatisticsQueryService) {
         List<ExperimentalFactor> result = new ArrayList<ExperimentalFactor>();
-        List<String> efs = atlasStatisticsQueryService.getScoringEfsForBioEntity(getGeneId(), UP_DOWN, ef);
-        efs.removeAll(omittedEfs);
+        List<EfAttribute> efs = atlasStatisticsQueryService.getScoringEfsForBioEntity(getGeneId(), UP_DOWN, ef);
+
+        Collections2.filter(efs,
```
Comment (alf239) [inline atlas-web/src/main/java/ae3/model/AtlasGene.java]: `Collections2.filter()` does not change the collection—it creates a new (lazy) one which is filtered. This line does not do anything.

Comment (rpetry) [inline atlas-web/src/main/java/ae3/model/AtlasGene.java]: How about now - c.f. commit/0d18d58c972090b1886df7aa3e040f0a4d55374f

**Comment by alf239** [comment]
```
Tests in error: 
  test_getGeneProperties(ae3.service.structuredquery.AtlasStructuredQueryServiceTest)
  test_doStructuredAtlasQuery(ae3.service.structuredquery.AtlasStructuredQueryServiceTest)
  test_getStats(ae3.service.structuredquery.AtlasStructuredQueryServiceTest)
  test_getExperimentCountsForGene(ae3.service.AtlasStatisticsQueryServiceTest)
  test_getStatisticsOrQuery(ae3.service.AtlasStatisticsQueryServiceTest)
  test_scoreQuery(ae3.service.AtlasStatisticsQueryServiceTest)
  test_getScoresAcrossAllEfos(ae3.service.AtlasStatisticsQueryServiceTest)
  getExperimentsForGeneAndAttribute(ae3.service.AtlasStatisticsQueryServiceTest)
  test_getSortedGenes(ae3.service.AtlasStatisticsQueryServiceTest)
  test_getScoringAttributesForGenes(ae3.service.AtlasStatisticsQueryServiceTest)
  test_getScoringExperimentsForGeneAndAttribute(ae3.service.AtlasStatisticsQueryServiceTest)
  test_getAttributesForEfo(ae3.service.AtlasStatisticsQueryServiceTest)
  test_getExperimentsSortedByPvalueTRank(ae3.service.AtlasStatisticsQueryServiceTest)
  test_getScoringEfsForGene(ae3.service.AtlasStatisticsQueryServiceTest)
  test_getScoringEfvsForGene(ae3.service.AtlasStatisticsQueryServiceTest)
  test_getExperimentsForGeneAndEf(ae3.service.AtlasStatisticsQueryServiceTest)

Tests run: 106, Failures: 4, Errors: 16, Skipped: 0
```

```
java.io.InvalidClassException: uk.ac.ebi.gxa.statistics.EfvAttribute; local class incompatible: stream classdesc serialVersionUID = -8117173650721973734, local class serialVersionUID = -4674735573391531209
```

**Comment by alf239** [comment]
and, well, _This pull request cannot be automatically merged._

**Comment by rpetry** [comment]
Latest comments addressed - thanks

**Pull request merged**